### PR TITLE
repository attributes

### PR DESCRIPTION
**Pull request by anandjakhaniya**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/entities/versioning/RepositoryEntity.java [hunk 1]
```diff
@@ -1,54 +1,58 @@
...
     this.date_created = new Date().getTime();
     this.date_updated = new Date().getTime();
-    if (repositoryVisibility != null) {
-      this.repository_visibility = repositoryVisibility.getNumber();
-    }
-    if (repositoryAccessModifier != null) {
-      this.repositoryAccessModifier = repositoryAccessModifier.getNumber();
-    }
+    this.repository_visibility = repository.getRepositoryVisibilityValue();
+    this.repositoryAccessModifier = repository.getRepositoryAccessModifierValue();
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/entities/versioning/RepositoryEntity.java]: not sure about these changes

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java [hunk 2]
```diff
@@ -62,5 +64,20 @@ BranchEntity getBranchEntity(Session session, Long repoId, String branchName)
...
+
+  void addRepositoryAttributes(RepositoryIdentification repositoryId, List<KeyValue> attributesList)
+      throws InvalidProtocolBufferException, ModelDBException;
+
+  void updateRepositoryAttribute(RepositoryIdentification repositoryId, KeyValue attribute)
+      throws InvalidProtocolBufferException, ModelDBException;
+
+  List<KeyValue> getRepositoryAttributes(
+      RepositoryIdentification repositoryId, List<String> attributeKeysList, boolean getAll)
+      throws InvalidProtocolBufferException, ModelDBException;
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java]: we do not need this

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java [hunk 3]
```diff
@@ -544,8 +549,9 @@ public Boolean deleteRepositories(List<String> repositoryIds, ExperimentRunDAO e
       ListRepositoriesRequest.Response.Builder builder =
           ListRepositoriesRequest.Response.newBuilder();
 
-      repositoryEntities.forEach(
-          repositoryEntity -> builder.addRepositories(repositoryEntity.toProto()));
+      for (RepositoryEntity repositoryEntity : repositoryEntities) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java:553]: why was this necessary?

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java [hunk 4]
```diff
@@ -956,4 +964,132 @@ public void deleteBranchByCommit(Session session, Long repoId, String commitHash
...
+    } catch (Exception ex) {
+      if (ModelDBUtils.needToRetry(ex)) {
+        updateRepositoryAttribute(repositoryId, attribute);
+      } else {
+        throw ex;
+      }
+    }
+  }
+
+  @Override
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java]: not required

File: backend/src/main/java/ai/verta/modeldb/versioning/VersioningServiceImpl.java [hunk 5]
```diff
@@ -663,4 +667,169 @@ public void findRepositoriesBlobs(
...
+      responseObserver.onCompleted();
+
+    } catch (Exception e) {
+      ModelDBUtils.observeError(
+          responseObserver, e, GetRepositoryAttributes.Response.getDefaultInstance());
+    }
+  }
+
+  @Override
+  public void deleteRepositoryAttributes(
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/VersioningServiceImpl.java]: not required

File: protos/protos/public/modeldb/versioning/VersioningService.proto [hunk 6]
```diff
@@ -522,6 +524,41 @@ message CommitMultipartVersionedBlobArtifact {
...
+
+message UpdateRepositoryAttribute {
+    RepositoryIdentification repository_id = 1;
+    common.KeyValue attribute = 2;
+
+    message Response {
+    }
+}
+
+message GetRepositoryAttributes {
```
Comment (ravishetye) [inline protos/protos/public/modeldb/versioning/VersioningService.proto]: not required

File: backend/src/main/java/ai/verta/modeldb/entities/versioning/RepositoryEntity.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/entities/versioning/RepositoryEntity.java]: Instead of sending each repository field as a parameter, I have sent a repository object as a parameter so it will never be null. it will have default value.

**Review by ravishetye — CHANGES_REQUESTED** [review]
not changes to Service and repository are required , as per the design doc UpdateDataset{Name,Description,Attributes}, addDatasetAttributes,setDatasetVisibility,setDatasetWorkspace will make a call to RepositoryDAO.setRepository.

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java]: I have replaced `getDatasetAttributes` endpoint with `getRepositoryAttributes` that's why this is added.

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java:553]: While converting Repository object to proto object it will throw `InvalidProtocolBufferException` exception dure to `keyvalue` conversion and we can not rethrow it from lambda so I have to change it to normal foreach. so yes it is necessary.

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAO.java]: Remove `getRepositoryAttributes` endpoint

File: backend/src/main/java/ai/verta/modeldb/versioning/VersioningServiceImpl.java (hunk 5 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/VersioningServiceImpl.java]: Remove `deleteRepositoryAttributes` endpoint

File: backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java]: Remove `getRepositoryAttributes` endpoint

File: protos/protos/public/modeldb/versioning/VersioningService.proto (hunk 6 above)
Comment (anandjakhaniya) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: Removed `getRepositoryAttributes` endpoint

**Comment by anandjakhaniya** [comment]
Removed all endpoints of repository attributes and added logic in the update repository endpoint for the (add, update) attributes.

**Review by ravishetye — APPROVED** [review]
